### PR TITLE
Expose dependency checks route

### LINKED ISSUE #53: delete dependency listing
endpoint that gives list of tablenames/entries that are dependent and whether user has the right to delete those tables. 
Screen Shot 2021-02-08 at 4 43 58 PM

### PR DESCRIPTION
**Pull request by guzman-raphael**
This PR introduces:
- `/api/record/dependency` - this route allows checking the dependency linkage of specific records. For details of inaccessible tables, it will only return the first inaccessible table. Current support includes `GET` verb and parameters are supplied via query parameters as follows:
  - schema_name
  - table_name
  - restriction (filters in JSON encoded in base64). Specifically these are to be only primary keys.
- Compatibility with DJ NEURO hosted users

## Example

If we wanted to get the dependencies of records matching the primary key `a_id=0` in a `TableA` table within the `deps` schema, here is how to do that.

### Request

Restriction would be:
```json
{
    "a_id": 0
}
```
`base64` encoding the string representation of this JSON would yield `eyJhX2lkIjogMH0=`

Therefore,

GET `/api/record/dependency?schema_name=deps&table_name=TableA&restriction=eyJhX2lkIjogMH0=`

### Response

```json
{
    "dependencies": [
        {"accessible": true, "count": 1, "schema": "deps", "table": "#table_a"},
        {"accessible": true, "count": 2, "schema": "deps", "table": "#table_b"},
        {"accessible": true, "count": 3, "schema": "deps", "table": "#table_c"},
        {"accessible": false, "schema": "deps_secret", "table": "#diff_table_b"}
    ]
}
```

Fix #53

### CODE REVIEW

File: dj_gui_api_server/DJGUIAPIServer.py [hunk 1]
```diff
@@ -219,6 +221,31 @@ def insert_tuple(jwt_payload: dict):
         return str(e), 500
 
 
+@app.route('/api/record/dependency', methods=['GET'])
+@protected_route
+def record_dependency(jwt_payload: dict) -> dict:
+    """
+    Route to insert record. Expects:
+        (html:GET:Authorization): Must include in format of: bearer <JWT-Token>
+        (html:query_params): {"schema_name": <schema_name>, "table_name": <table_name>,
```
Comment (mahos) [inline dj_gui_api_server/DJGUIAPIServer.py:230]:
```suggestion
        (html:query_params): {"schemaName": <schema_name>, "tableName": <table_name>,
```
maybe change to match the format that's used on the frontside?

File: dj_gui_api_server/DJGUIAPIServer.py [hunk 2]
```diff
@@ -219,6 +221,31 @@ def insert_tuple(jwt_payload: dict):
...
+    :param jwt_payload: Dictionary containing databaseAddress, username and password
+        strings.
+    :type jwt_payload: dict
+    :return: If sucessfuly sends back a list of dependencies otherwise returns error
+    :rtype: dict
+    """
+    # Get dependencies
+    try:
+        dependencies = DJConnector.record_dependency(
+            jwt_payload, request.args.get('schema_name'), request.args.get('table_name'),
```
Comment (mahos) [inline dj_gui_api_server/DJGUIAPIServer.py:242]:
```suggestion
            jwt_payload, request.args.get('schemaName'), request.args.get('tableName'),
```
same as above - to match query param format from the frontside

**Review by mahos — COMMENTED** [review]
The list of dependencies is working nicely on the test branch - suggesting a query format change just for consistency sakes from the frontend side.

**Comment by guzman-raphael** [comment]
@mahos Thanks for the feedback. I've adjusted it and it is ready for review.

File: tests/test_dependencies.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    REST_dependencies = client.get(
+        f"""/api/record/dependency?schemaName={
+            schema_name}&tableName={table_name}&restriction={restriction}""",
+        headers=dict(Authorization=f'Bearer {underprivileged_token}')).json['dependencies']
+    REST_records = client.post('/api/fetch_tuples',
+                               headers=dict(Authorization=f'Bearer {underprivileged_token}'),
+                               json=dict(schemaName=schema_name,
+                                         tableName=table_name)).json['tuples']
+    # print(REST_dependencies)
+    # print(REST_records)
```
Comment (Synicix) [inline tests/test_dependencies.py:104]: Remove debug statements?

File: tests/test_dependencies.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    REST_dependencies = client.get(
+        f"""/api/record/dependency?schemaName={
+            schema_name}&tableName={table_name}&restriction={restriction}""",
+        headers=dict(Authorization=f'Bearer {token}')).json['dependencies']
+    REST_records = client.post('/api/fetch_tuples',
+                               headers=dict(Authorization=f'Bearer {token}'),
+                               json=dict(schemaName=schema_name,
+                                         tableName=table_name)).json['tuples']
+    # print(REST_dependencies)
+    # print(REST_records)
```
Comment (Synicix) [inline tests/test_dependencies.py:134]: Remove debug statements?

**Review by Synicix — CHANGES_REQUESTED** [review]
Looks, good, just a few minor stuff that needs to be removed

File: tests/test_dependencies.py (hunk 3 above)
Comment (guzman-raphael) [reply tests/test_dependencies.py]: Good catch. Addressed.

File: tests/test_dependencies.py (hunk 4 above)
Comment (guzman-raphael) [reply tests/test_dependencies.py]: Addressed.

**Review by Synicix — APPROVED** [review]
Looks good, merging

**Pull request merged**